### PR TITLE
Prevent the models returned from find_all to be modified in the middle of modifying

### PR DESCRIPTION
**Pull request by harris**
- do not execute the migration while collecting the input and output documents. 
- async gather so each batch could potentially run in parallel and not block each other. 
- verified it properly fixed the issues of any collections with size larger than 10,000 records.

### CODE REVIEW

File: beanie/migrations/controllers/iterative.py [hunk 1]
```diff
@@ -102,19 +104,22 @@ async def run(self, session):
...
-                    output_dict
-                )
+                    output_dict)
                 output_documents.append(output_document)
 
                 if len(output_documents) == self.batch_size:
-                    await self.output_document_model.replace_many(
-                        documents=output_documents
+                    all_migration_ops.append(
+                        self.output_document_model.replace_many(
```
Comment (jspiro) [inline beanie/migrations/controllers/iterative.py:112]: Suggest we wrap this promise with a print statement for each batch.

File: beanie/migrations/controllers/iterative.py [hunk 2]
```diff
@@ -102,19 +104,22 @@ async def run(self, session):
...
+                            documents=output_documents
+                        )
                     )
                     output_documents = []
 
             if output_documents:
-                await self.output_document_model.replace_many(
-                    documents=output_documents
+                all_migration_ops.append(
+                    self.output_document_model.replace_many(
```
Comment (jspiro) [inline beanie/migrations/controllers/iterative.py:120]: Same.

File: beanie/migrations/controllers/iterative.py [hunk 3]
```diff
@@ -102,19 +104,22 @@ async def run(self, session):
...
+                        documents=output_documents)
                 )
+            await asyncio.gather(*all_migration_ops)
```
Comment (jspiro) [inline beanie/migrations/controllers/iterative.py:123]: Or we can do it here and catch/rethrow errors. E.g.,
```
for op in asyncio.as_completed(all_migration_ops):
    try:
        print('Progress...')
        results = await op
    except Exception as e:
        print('Error:', e)
        throw e
```

**Comment by jspiro** [comment]
I'd rename this to `prevent migrations from never completing`

**Comment by harris** [comment]
@roman-right do you think you could take a look at this?  We actually had a data loss due to this bug in migration.

**Comment by roman-right** [comment]
Merged. Thank you, guys!
It will be published in the next beta or normal release today or tomorrow (it depends on how many PRs will be included)

**Pull request merged**